File: analytics/extract.py
#!/usr/bin/env python3
"""
extract.py — Single-pass MT5 report parser.

Reads MT5 backtest report (.htm or .htm.xml / SpreadsheetML) and produces:
  - metrics.json  (aggregate summary)
  - deals.csv     (all deals, 13 columns)
  - deals.json    (deals as JSON array)

Usage:
    python3 analytics/extract.py report.htm --output-dir reports/20250101_123456/
"""

import argparse
import csv
import json
import os
import re
import sys
import xml.etree.ElementTree as ET
from pathlib import Path
from typing import Optional


# MT5 backtest report deals table columns (actual order from HTML):
# Time, Deal, Symbol, Type, Direction, Volume, Price, Order, Commission, Swap, Profit, Balance, Comment
DEAL_COLUMNS = [
    "time", "deal", "symbol", "type", "entry", "volume", "price",
    "order", "commission", "swap", "profit", "balance", "comment"
]


def detect_format(path: str) -> str:
    """Return 'xml' for SpreadsheetML, 'html' for legacy HTML report."""
    if path.endswith('.xml') or path.endswith('.htm.xml'):
        return 'xml'
    # Peek at file header
    with open(path, 'rb') as f:
        header = f.read(512)
    if b'<?xml' in header or b'Workbook' in header:
        return 'xml'
    return 'html'


def read_text(path: str) -> str:
    """Read file, handling UTF-16 (MT5 default) and latin-1 fallback."""
    with open(path, 'rb') as f:
        raw = f.read()
    for encoding in ('utf-16', 'utf-8', 'latin-1'):
        try:
            return raw.decode(encoding)
        except (UnicodeDecodeError, LookupError):
            continue
    return raw.decode('latin-1', errors='replace')


def strip_tags(html: str) -> str:
    return re.sub(r'<[^>]+>', '', html).strip()


# ── HTML parser ───────────────────────────────────────────────────────────────

def parse_html(path: str) -> tuple[dict, list[dict]]:
    text = read_text(path)

    metrics = _parse_metrics_html(text)
    deals = _parse_deals_html(text)
    return metrics, deals


def _parse_metrics_html(text: str) -> dict:
    """Extract aggregate metrics from the summary table."""
    m = {}

    # MT5 report HTML format: MetricLabel:</td>\r\n<td nowrap><b>VALUE</b></td>
    # Helper patterns — values always wrapped in <b>...</b>
    _b = r'[^<]*</td>\s*<td[^>]*>\s*<b>([-\d\s.,]+)</b>'           # plain number
    _b_pct = r'[^<]*</td>\s*<td[^>]*>\s*<b>[^(]*\(([\d.,]+)%\)'    # "abs (pct%)" — capture pct
    patterns = {
        'net_profit':      r'Net\s+Profit' + _b,
        'profit_factor':   r'Profit\s+Factor' + _b,
        'max_dd_pct':      r'Equity\s+Drawdown\s+Maximal' + _b_pct,
        'sharpe_ratio':    r'Sharpe\s+Ratio' + _b,
        'total_trades':    r'Total\s+Trades' + _b,
        'recovery_factor': r'Recovery\s+Factor' + _b,
        'win_rate_pct':    r'Profit\s+Trades\s+\(%' + _b_pct,
        'gross_profit':    r'Gross\s+Profit' + _b,
        'gross_loss':      r'Gross\s+Loss' + _b,
    }

    for key, pattern in patterns.items():
        match = re.search(pattern, text, re.IGNORECASE | re.DOTALL)
        if match:
            val = match.group(1).replace(' ', '').replace(',', '').strip()
            try:
                m[key] = float(val)
            except ValueError:
                pass

    # Trades needs int
    if 'total_trades' in m:
        m['total_trades'] = int(m['total_trades'])

    return m


def _parse_deals_html(text: str) -> list[dict]:
    """Extract deal rows from the deals table."""
    # Find deals section (after "Deals" header)
    deals_section = re.search(
        r'<tr[^>]*>.*?Deal.*?Time.*?Type.*?Direction.*?Volume.*?</tr>(.*)',
        text, re.DOTALL | re.IGNORECASE
    )
    if not deals_section:
        return []

    rows = re.findall(
        r'<tr[^>]*>(.*?)</tr>',
        deals_section.group(1),
        re.DOTALL | re.IGNORECASE
    )

    deals = []
    for row in rows:
        cells = re.findall(r'<td[^>]*>(.*?)</td>', row, re.DOTALL | re.IGNORECASE)
        cells = [strip_tags(c).replace(',', '') for c in cells]

        if len(cells) < 3 or not cells[0]:
            continue
        # Skip balance/deposit/credit rows — 'balance' appears in the Type column (index 3)
        # or sometimes in index 1; check first 5 cells
        if any(c.strip().lower() in ('balance', 'credit') for c in cells[:5]):
            continue

        deal = {}
        for i, col in enumerate(DEAL_COLUMNS):
            deal[col] = cells[i] if i < len(cells) else ''
        deals.append(deal)

    return deals


# ── XML parser (SpreadsheetML) ─────────────────────────────────────────────────

def parse_xml(path: str) -> tuple[dict, list[dict]]:
    """Parse MT5 SpreadsheetML optimization/report XML."""
    tree = ET.parse(path)
    root = tree.getroot()

    # Namespace handling — MT5 XML uses Excel namespace
    ns = {}
    ns_match = re.match(r'\{([^}]+)\}', root.tag)
    if ns_match:
        ns['ss'] = ns_match.group(1)

    def tag(name):
        return f"{{{ns['ss']}}}{name}" if ns else name

    metrics = {}
    deals = []
    in_deals_sheet = False

    for sheet in root.iter(tag('Worksheet')):
        sheet_name = sheet.get(f"{{{ns['ss']}}}Name" if ns else 'Name', '')

        if 'result' in sheet_name.lower() or 'report' in sheet_name.lower():
            metrics = _parse_metrics_xml(sheet, tag)
        elif 'deal' in sheet_name.lower() or 'trade' in sheet_name.lower():
            deals = _parse_deals_xml(sheet, tag)
        elif sheet_name == '':
            # Unnamed sheet — check if it has deal-like structure
            rows = list(sheet.iter(tag('Row')))
            if len(rows) > 5:
                # Try to parse as deals
                candidate = _parse_deals_xml(sheet, tag)
                if candidate:
                    deals = candidate

    return metrics, deals


def _cell_value(cell, tag) -> str:
    data = cell.find(tag('Data'))
    return data.text.strip() if data is not None and data.text else ''


def _parse_metrics_xml(sheet, tag) -> dict:
    m = {}
    for row in sheet.iter(tag('Row')):
        cells = [_cell_value(c, tag) for c in row.iter(tag('Cell'))]
        if len(cells) < 2:
            continue
        key = cells[0].lower()
        val = cells[1].replace(',', '').strip()
        try:
            fval = float(val)
            if 'net profit' in key or 'net_profit' in key:
                m['net_profit'] = fval
            elif 'profit factor' in key:
                m['profit_factor'] = fval
            elif 'drawdown' in key and '%' in cells[1]:
                m['max_dd_pct'] = fval
            elif 'sharpe' in key:
                m['sharpe_ratio'] = fval
            elif 'total trades' in key:
                m['total_trades'] = int(fval)
        except (ValueError, AttributeError):
            pass
    return m


def _parse_deals_xml(sheet, tag) -> list[dict]:
    deals = []
    header_found = False
    col_map = {}

    for row in sheet.iter(tag('Row')):
        cells = [_cell_value(c, tag) for c in row.iter(tag('Cell'))]

        if not header_found:
            # Detect header row
            if any(h in str(cells).lower() for h in ('time', 'type', 'volume', 'profit')):
                header_found = True
                for i, h in enumerate(cells):
                    h_lower = h.lower().strip()
                    for col in DEAL_COLUMNS:
                        if col in h_lower or h_lower in col:
                            col_map[i] = col
                            break
            continue

        if not cells or not cells[0]:
            continue

        deal = {}
        for i, val in enumerate(cells):
            col = col_map.get(i)
            if col:
                deal[col] = val.replace(',', '')

        if deal:
            deals.append(deal)

    return deals


# ── Writer ────────────────────────────────────────────────────────────────────

def write_outputs(metrics: dict, deals: list[dict], output_dir: str) -> dict:
    os.makedirs(output_dir, exist_ok=True)

    metrics_path = os.path.join(output_dir, 'metrics.json')
    deals_csv_path = os.path.join(output_dir, 'deals.csv')
    deals_json_path = os.path.join(output_dir, 'deals.json')

    with open(metrics_path, 'w') as f:
        json.dump(metrics, f, indent=2)

    with open(deals_json_path, 'w') as f:
        json.dump(deals, f, indent=2)

    if deals:
        all_keys = DEAL_COLUMNS
        with open(deals_csv_path, 'w', newline='') as f:
            writer = csv.DictWriter(f, fieldnames=all_keys, extrasaction='ignore')
            writer.writeheader()
            writer.writerows(deals)
    else:
        # Write empty CSV with headers
        with open(deals_csv_path, 'w', newline='') as f:
            writer = csv.writer(f)
            writer.writerow(DEAL_COLUMNS)

    return {
        'metrics': metrics_path,
        'deals_csv': deals_csv_path,
        'deals_json': deals_json_path,
    }


# ── Main ──────────────────────────────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser(description='Extract MT5 backtest report')
    parser.add_argument('report', help='Path to report.htm or report.htm.xml')
    parser.add_argument('--output-dir', default='.', help='Output directory')
    parser.add_argument('--stdout', action='store_true',
                        help='Print metrics JSON to stdout instead of writing files')
    args = parser.parse_args()

    fmt = detect_format(args.report)

    if fmt == 'xml':
        metrics, deals = parse_xml(args.report)
    else:
        metrics, deals = parse_html(args.report)

    if not metrics:
        print(f"WARNING: No aggregate metrics found in report", file=sys.stderr)
    if not deals:
        print(f"WARNING: No deals found in report (check date range and symbol)", file=sys.stderr)

    if args.stdout:
        json.dump({'metrics': metrics, 'deals_count': len(deals)}, sys.stdout, indent=2)
        print()
        return

    paths = write_outputs(metrics, deals, args.output_dir)

    print(f"Extracted: {len(deals)} deals, {len(metrics)} metrics")
    for name, path in paths.items():
        print(f"  {name}: {path}")


if __name__ == '__main__':
    main()

File: src/analytics/extract.rs
use anyhow::{anyhow, Result};
use std::collections::HashMap;
use std::fs::{self, File};
use std::io::Write;
use std::path::{Path, PathBuf};

use crate::models::{Deal, Metrics};

pub struct ReportExtractor;

impl ReportExtractor {
    pub fn new() -> Self {
        Self
    }

    pub fn extract(&self, report_path: &str, output_dir: &str) -> Result<ExtractionResult> {
        let format = Self::detect_format(report_path);
        
        let (metrics, deals) = match format {
            ReportFormat::Xml => self.parse_xml(report_path)?,
            ReportFormat::Html => self.parse_html(report_path)?,
        };

        fs::create_dir_all(output_dir)?;

        let metrics_path = Path::new(output_dir).join("metrics.json");
        let deals_csv_path = Path::new(output_dir).join("deals.csv");
        let deals_json_path = Path::new(output_dir).join("deals.json");

        self.write_metrics(&metrics, &metrics_path)?;
        self.write_deals_json(&deals, &deals_json_path)?;
        self.write_deals_csv(&deals, &deals_csv_path)?;

        Ok(ExtractionResult {
            metrics,
            deals,
            metrics_path,
            deals_csv_path,
            deals_json_path,
        })
    }

    fn detect_format(path: &str) -> ReportFormat {
        if path.ends_with(".xml") || path.ends_with(".htm.xml") {
            return ReportFormat::Xml;
        }

        if let Ok(file) = fs::read(path) {
            let header = &file[..file.len().min(512)];
            if header.windows(5).any(|w| w == b"<?xml") || header.windows(8).any(|w| w == b"Workbook") {
                return ReportFormat::Xml;
            }
        }

        ReportFormat::Html
    }

    fn parse_html(&self, path: &str) -> Result<(Metrics, Vec<Deal>)> {
        let text = Self::read_text(path)?;
        
        let metrics = Metrics::from_html(&text)
            .ok_or_else(|| anyhow!("No metrics found in HTML report"))?;
        
        let deals = self.parse_deals_html(&text)?;

        Ok((metrics, deals))
    }

    fn parse_xml(&self, path: &str) -> Result<(Metrics, Vec<Deal>)> {
        let text = Self::read_text(path)?;
        
        let metrics = Metrics::from_html(&text)
            .unwrap_or_default();
        
        let deals = self.parse_deals_xml(&text)?;

        Ok((metrics, deals))
    }

    fn parse_deals_html(&self, text: &str) -> Result<Vec<Deal>> {
        let mut deals = Vec::new();

        let re = regex::Regex::new(r"<tr[^>]*>.*?Deal.*?Time.*?Type.*?Direction.*?</tr>(.*)")
            .map_err(|e| anyhow!("Regex error: {}", e))?;

        if let Some(captures) = re.captures(text) {
            let section = captures.get(1).map(|m| m.as_str()).unwrap_or("");

            let row_re = regex::Regex::new(r"<tr[^>]*>(.*?)</tr>")
                .map_err(|e| anyhow!("Regex error: {}", e))?;

            for row_caps in row_re.captures_iter(section) {
                let row = row_caps.get(1).map(|m| m.as_str()).unwrap_or("");
                
                let cell_re = regex::Regex::new(r"<td[^>]*>(.*?)</td>")
                    .map_err(|e| anyhow!("Regex error: {}", e))?;
                
                let cells: Vec<String> = cell_re.captures_iter(row)
                    .filter_map(|cap| cap.get(1))
                    .map(|m| Self::strip_tags(m.as_str()))
                    .map(|s| s.replace(',', ""))
                    .collect();

                if cells.len() < 3 || cells[0].is_empty() {
                    continue;
                }

                if cells.iter().take(5).any(|c| {
                    let c_lower = c.trim().to_lowercase();
                    c_lower == "balance" || c_lower == "credit"
                }) {
                    continue;
                }

                let deal = Deal {
                    time: cells.get(0).cloned().unwrap_or_default(),
                    deal: cells.get(1).cloned().unwrap_or_default(),
                    symbol: cells.get(2).cloned().unwrap_or_default(),
                    deal_type: cells.get(3).cloned().unwrap_or_default(),
                    entry: cells.get(4).cloned().unwrap_or_default(),
                    volume: cells.get(5).and_then(|s| s.parse().ok()).unwrap_or(0.0),
                    price: cells.get(6).and_then(|s| s.parse().ok()).unwrap_or(0.0),
                    order: cells.get(7).cloned().unwrap_or_default(),
                    commission: cells.get(8).and_then(|s| s.parse().ok()).unwrap_or(0.0),
                    swap: cells.get(9).and_then(|s| s.parse().ok()).unwrap_or(0.0),
                    profit: cells.get(10).and_then(|s| s.parse().ok()).unwrap_or(0.0),
                    balance: cells.get(11).and_then(|s| s.parse().ok()).unwrap_or(0.0),
                    comment: cells.get(12).cloned().unwrap_or_default(),
                    magic: cells.get(13).cloned(),
                };

                deals.push(deal);
            }
        }

        Ok(deals)
    }

    fn parse_deals_xml(&self, text: &str) -> Result<Vec<Deal>> {
        let mut deals = Vec::new();
        let mut header_found = false;
        let mut col_map: HashMap<usize, String> = HashMap::new();

        let row_re = regex::Regex::new(r"<Row[^>]*>(.*?)</Row>")
            .map_err(|e| anyhow!("Regex error: {}", e))?;
        
        let cell_re = regex::Regex::new(r"<Cell[^>]*>.*?<Data[^>]*>(.*?)</Data>.*?</Cell>")
            .map_err(|e| anyhow!("Regex error: {}", e))?;

        for row_caps in row_re.captures_iter(text) {
            let row = row_caps.get(1).map(|m| m.as_str()).unwrap_or("");
            
            let cells: Vec<String> = cell_re.captures_iter(row)
                .filter_map(|cap| cap.get(1))
                .map(|m| Self::strip_tags(m.as_str()).replace(',', ""))
                .collect();

            if !header_found {
                let row_str = cells.join("").to_lowercase();
                if row_str.contains("time") || row_str.contains("type") || row_str.contains("volume") {
                    header_found = true;
                    for (i, h) in cells.iter().enumerate() {
                        let h_lower = h.to_lowercase().trim().to_string();
                        let deal_columns = ["time", "deal", "symbol", "type", "entry", "volume", "price", "order", "commission", "swap", "profit", "balance", "comment"];
                        for col in &deal_columns {
                            if h_lower.contains(col) || col.contains(&h_lower) {
                                col_map.insert(i, col.to_string());
                                break;
                            }
                        }
                    }
                    continue;
                }
            }

            if cells.is_empty() || cells[0].is_empty() {
                continue;
            }

            let mut deal_map: HashMap<String, String> = HashMap::new();
            for (i, val) in cells.iter().enumerate() {
                if let Some(col) = col_map.get(&i) {
                    deal_map.insert(col.clone(), val.clone());
                }
            }

            if !deal_map.is_empty() {
                let deal = Deal {
                    time: deal_map.get("time").cloned().unwrap_or_default(),
                    deal: deal_map.get("deal").cloned().unwrap_or_default(),
                    symbol: deal_map.get("symbol").cloned().unwrap_or_default(),
                    deal_type: deal_map.get("type").cloned().unwrap_or_default(),
                    entry: deal_map.get("entry").cloned().unwrap_or_default(),
                    volume: deal_map.get("volume").and_then(|s| s.parse().ok()).unwrap_or(0.0),
                    price: deal_map.get("price").and_then(|s| s.parse().ok()).unwrap_or(0.0),
                    order: deal_map.get("order").cloned().unwrap_or_default(),
                    commission: deal_map.get("commission").and_then(|s| s.parse().ok()).unwrap_or(0.0),
                    swap: deal_map.get("swap").and_then(|s| s.parse().ok()).unwrap_or(0.0),
                    profit: deal_map.get("profit").and_then(|s| s.parse().ok()).unwrap_or(0.0),
                    balance: deal_map.get("balance").and_then(|s| s.parse().ok()).unwrap_or(0.0),
                    comment: deal_map.get("comment").cloned().unwrap_or_default(),
                    magic: deal_map.get("magic").cloned(),
                };
                deals.push(deal);
            }
        }

        Ok(deals)
    }

    fn write_metrics(&self, metrics: &Metrics, path: &Path) -> Result<()> {
        let json = serde_json::to_string_pretty(metrics)?;
        let mut file = File::create(path)?;
        file.write_all(json.as_bytes())?;
        Ok(())
    }

    fn write_deals_json(&self, deals: &[Deal], path: &Path) -> Result<()> {
        let json = serde_json::to_string_pretty(deals)?;
        let mut file = File::create(path)?;
        file.write_all(json.as_bytes())?;
        Ok(())
    }

    fn write_deals_csv(&self, deals: &[Deal], path: &Path) -> Result<()> {
        let mut file = File::create(path)?;
        writeln!(file, "time,deal,symbol,type,entry,volume,price,order,commission,swap,profit,balance,comment")?;
        
        for deal in deals {
            writeln!(file, "{},{},{},{},{},{},{},{},{},{},{},{},\"{}\"",
                deal.time,
                deal.deal,
                deal.symbol,
                deal.deal_type,
                deal.entry,
                deal.volume,
                deal.price,
                deal.order,
                deal.commission,
                deal.swap,
                deal.profit,
                deal.balance,
                deal.comment.replace('"', "\"\"")
            )?;
        }
        
        Ok(())
    }

    fn read_text(path: &str) -> Result<String> {
        let raw = fs::read(path)?;
        
        if raw.starts_with(&[0xFF, 0xFE]) || raw.starts_with(&[0xFE, 0xFF]) {
            // UTF-16 BOM
            let text = String::from_utf16_lossy(
                raw.chunks_exact(2)
                    .map(|c| u16::from_le_bytes([c[0], c[1]]))
                    .collect::<Vec<_>>()
                    .as_slice()
            );
            return Ok(text);
        }
        
        if let Ok(text) = String::from_utf8(raw.clone()) {
            return Ok(text);
        }
        
        Ok(String::from_utf8_lossy(&raw).to_string())
    }

    fn strip_tags(html: &str) -> String {
        let re = regex::Regex::new(r"<[^>]+>").unwrap();
        re.replace_all(html, "").trim().to_string()
    }
}

pub struct ExtractionResult {
    pub metrics: Metrics,
    pub deals: Vec<Deal>,
    pub metrics_path: PathBuf,
    pub deals_csv_path: PathBuf,
    pub deals_json_path: PathBuf,
}

#[derive(Debug, Clone, Copy)]
enum ReportFormat {
    Html,
    Xml,
}
